### PR TITLE
Review workflow

### PR DESCRIPTION
**Pull request by MaximMoinat**
New features:
* Flag status (from map-variables branch)
* Provenance, author and timestamp (from map-variables branch)
* Equivalence
* Random reviewer assignment
* Minor other updates to support review workflow

Todo:
- [x] Give list of reviewers to the reviewer assignment
- [x] Evenly divide codes between reviewers
- [x] ~Filtering of code list by reviewer~

### CODE REVIEW

**Review by blootsvoets — COMMENTED** [review]
Fine in general. I have some questions about the data classes and the row mapper. Other than that I didn't test the UI.

File: src/org/ohdsi/usagi/ui/MappingDetailPanel.java [hunk 1]
```diff
@@ -431,17 +442,11 @@ public void replaceConcepts(Concept concept) {
...
-		Collections.sort(rows, new Comparator<Integer>() {
-
-			@Override
-			public int compare(Integer o1, Integer o2) {
-				return o2.compareTo(o1);
-			}
-		});
+		rows.sort(Comparator.reverseOrder());
 		for (int row : rows)
 			codeMapping.targetConcepts.remove(row);
```
[the remark this answers is not in the document]
Comment (MaximMoinat) [reply src/org/ohdsi/usagi/ui/MappingDetailPanel.java:451]: Done. Note: sorting ints in reversed order using a stream requires boxing to Integer
https://stackoverflow.com/questions/30740471/how-to-sort-an-intstream-in-reverse-order

File: src/org/ohdsi/usagi/ui/MappingTablePanel.java [hunk 2]
```diff
@@ -234,11 +244,26 @@ public Object getValueAt(int row, int col) {
 		}
 
 		public boolean isCellEditable(int row, int col) {
+			col = resolveColumnIndex(col);
+			if (col == 20) {
```
[the remark this answers is not in the document]
Comment (MaximMoinat) [reply src/org/ohdsi/usagi/ui/MappingTablePanel.java:248]: Indeed, and you have discovered a bug as well. Thanks!

File: src/org/ohdsi/usagi/ui/TargetConceptTableModel.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import javax.swing.table.AbstractTableModel;
+import java.util.ArrayList;
+import java.util.List;
+
+class TargetConceptTableModel extends AbstractTableModel {
+    private static final long serialVersionUID = -4978479688021056281L;
+
+    private final String termColumnName = "Term";
+    private final String[] columnNames = {"Concept ID", "Concept name", "Domain", "Concept class", "Vocabulary", "Concept code",
```
[the remark this answers is not in the document]
Comment (MaximMoinat) [reply src/org/ohdsi/usagi/ui/TargetConceptTableModel.java:13]: Agree that this whole column system needs a complete overhaul. However, not worth the effort at the moment.

File: src/org/ohdsi/usagi/ReadCodeMappingsFromFile.java [hunk 4]
```diff
@@ -58,20 +59,35 @@ private void readNext() {
...
+
 				try {
 					buffer.comment = row.get("comment");
 				} catch (Exception e) {
 					buffer.comment = "";
 				}
-				while (row != null && new SourceCode(row).sourceCode.equals(buffer.sourceCode.sourceCode)
+				while (row != null
+						&& new SourceCode(row).sourceCode.equals(buffer.sourceCode.sourceCode)
 						&& new SourceCode(row).sourceName.equals(buffer.sourceCode.sourceName)) {
```
[the remark this answers is not in the document]
Comment (MaximMoinat) [reply src/org/ohdsi/usagi/ReadCodeMappingsFromFile.java]: It is overcomplicated indeed. However, it works and not worth the effort to refactor.

**Review by blootsvoets — APPROVED** [review]

**Comment by MaximMoinat** [comment]
Additional features to be added after first round of testing:
- [ ] Set equivalence information also when flagging a mapping
- [ ] Reset equivalence to `UNREVIEWED` when unapproved or unflagged
- [ ] Assign reviewers only to selected rows.
- [ ] (optional) Add in mapping types for target concepts: `EVENT`, `VALUE` and `UNIT` for "Maps to", "Maps to value" and "Maps to unit" respectively. `EVENT` is the default type.

@mik-ohdsi Sounds good?

**Pull request merged**